### PR TITLE
Parse X509 Requests

### PR DESCRIPTION
**Pull request by reaperhulk**
Adds limited X509 request parsing. No extensions, no version (do we even care?), no verification.

Depends on #1697

### CODE REVIEW

**Comment by alex** [comment]
Tests are failing because stuff refers to the old `interfaces.` modules

**Comment by alex** [comment]
In my experience these are almost always called CSRs, does that match your experience?

**Comment by reaperhulk** [comment]
Yes you typically see them called CSRs (and I did use `R509::CSR` in ruby)

File: docs/x509.rst [hunk 1]
```diff
@@ -77,6 +77,58 @@ Loading Certificates
     >>> cert.serial
     2
 
+Loading Certificate Requests
```
Comment (public) [inline docs/x509.rst]: We use sentence case for titles most of the time.

**Comment by alex** [comment]
Haven't reviewed closely at all, but the version numbers need to be bumped.

**Comment by alex** [comment]
retest this please

File: CHANGELOG.rst [hunk 2]
```diff
@@ -12,6 +12,8 @@ Changelog
   :class:`~cryptography.hazmat.primitives.asymmetric.dsa.DSAPublicKeyWithSerialization`,
   and
   :class:`~cryptography.hazmat.primitives.asymmetric.ec.EllipticCurvePublicKeyWithSerialization`.
+* Add support for parsing X.509 requests with
```
Comment (alex) [inline CHANGELOG.rst]: bare requests here

File: docs/x509.rst [hunk 3]
```diff
@@ -77,6 +77,58 @@ Loading Certificates
...
     2
 
+Loading CSRs
+~~~~~~~~~~~~
+
+.. function:: load_pem_x509_csr(data, backend)
+
+    .. versionadded:: 0.9
+
+    Deserialize a certificate request from PEM encoded data. PEM requests are
```
Comment (alex) [inline docs/x509.rst]: "signing" here

File: docs/x509.rst [hunk 4]
```diff
@@ -77,6 +77,58 @@ Loading Certificates
...
+    Ha9urDAJ1wIDAQABoDUwMwYJKoZIhvcNAQkOMSYwJDAiBgNVHREEGzAZggl3b3Js
+    ZC5jb22CDHdoYXRldmVyLmNvbTANBgkqhkiG9w0BAQUFAAOCAQEAS4Ro6h+z52SK
+    YSLCYARpnEu/rmh4jdqndt8naqcNb6uLx9mlKZ2W9on9XDjnSdQD9q+ZP5aZfESw
+    R0+rJhW9ZrNa/g1pt6M24ihclHYDAxYMWxT1z/TXXGM3TmZZ6gfYlNE1kkBuODHa
+    UYsR/1Ht1E1EsmmUimt2n+zQR2K8T9Coa+boaUW/GsTEuz1aaJAkj5ZvTDiIhRG4
+    AOCqFZOLAQmCCNgJnnspD9hDz/Ons085LF5wnYjN4/Nsk5tS6AGs3xjZ3jPoOGGn
+    82WQ9m4dBGoVDZXsobVTaN592JEYwN5iu72zRn7Einb4V4H5y3yD2dD4yWPlt4pk
+    5wFkeYsZEA==
+    -----END CERTIFICATE REQUEST-----
+    """.strip()
```
Comment (alex) [inline docs/x509.rst]: Should we put this at the top of the document? It's a bit disruptive when you're reading through the plaintext.

File: docs/x509.rst [hunk 5]
```diff
@@ -196,6 +248,49 @@ X.509 Certificate Object
             >>> isinstance(cert.signature_hash_algorithm, hashes.SHA256)
             True
 
+X.509 CSR Object
```
Comment (alex) [inline docs/x509.rst]: I think we should use the full "Certificate Signing Request" here

File: src/cryptography/hazmat/backends/multibackend.py [hunk 6]
```diff
@@ -328,3 +328,14 @@ def load_der_x509_certificate(self, data):
             "This backend does not support X.509.",
             _Reasons.UNSUPPORTED_X509
         )
+
+    def load_pem_x509_csr(self, data):
+        for b in self._filtered_backends(
+            X509Backend
+        ):
```
Comment (alex) [inline src/cryptography/hazmat/backends/multibackend.py]: This can go on one line.

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 7]
```diff
@@ -20,6 +20,43 @@
...
+        )
+        value = backend._ffi.buffer(buf[0], res)[:].decode('utf8')
+        oid = _obj2txt(backend, obj)
+        attributes.append(
+            x509.NameAttribute(
+                x509.ObjectIdentifier(oid), value
+            )
+        )
+
+    return x509.Name(attributes)
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Can we move these two refactorings into their own PR?

**Comment by alex** [comment]
Not sure how I feel about naming the class "CSR"

**Comment by reaperhulk** [comment]
I'm fine with changing it to CertificateSigningRequest. I've been doing PKI too long to know what's intuitive or understandable so I'm happy to let others make that call ;)

**Comment by alex** [comment]
I'm not sure; we use initialisms in a few places, but mostly it's where
things are primarily known that way (e.g. AES or PEM), I'm genuinely not
sure which is more clear here.

On Wed, Mar 11, 2015 at 9:40 PM, Paul Kehrer <notifications@github.com>
wrote:

> I'm fine with changing it to CertificateSigningRequest. I've been doing
> PKI too long to know what's intuitive or understandable so I'm happy to let
> others make that call ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1702#issuecomment-78409026>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

**Comment by alex** [comment]
Needs a merge/rebase

**Comment by alex** [comment]
a) Needs a rebase
b) Mostly looks good, I haven't closely checked the tests for completeness or that they match the files though
c) We need to figure out the naming issue and make the docs, etc. consistent before we can advance

File: docs/x509.rst (hunk 4 above)
Comment (reaperhulk) [inline docs/x509.rst]: This has now been moved to the top.

File: docs/hazmat/backends/interfaces.rst [hunk 8]
```diff
@@ -509,3 +509,12 @@ A specific ``backend`` may provide one or more of these interfaces.
         :param bytes data: DER formatted certificate data.
 
         :returns: An instance of :class:`~cryptography.x509.Certificate`.
+
+    .. method:: load_pem_x509_csr(data)
+
+        .. versionadded:: 0.9
+
+        :param bytes data: PEM formatted certificate request data.
```
Comment (alex) [inline docs/hazmat/backends/interfaces.rst]: Missing "signing" here, should we change the method name as well?

File: docs/x509.rst (hunk 4 above)
Comment (alex) [inline docs/x509.rst]: When I said the top, I meant all the way at the top

File: docs/hazmat/backends/interfaces.rst (hunk 8 above)
Comment (reaperhulk) [inline docs/hazmat/backends/interfaces.rst]: I discussed this with glyph a bit on IRC and it was his contention that `load_pem_x509_csr` is better than `load_pem_x509_certificate_signing_request` since pem is already an abbreviation. I am leaning towards leaving it as csr here but could be persuaded otherwise. `load_pem_x509_certificate_signing_request` is a pretty long name though.

File: docs/x509.rst (hunk 4 above)
Comment (reaperhulk) [inline docs/x509.rst]: Should I move the certificate PEM setup to live with it as part of this PR?

File: docs/hazmat/backends/interfaces.rst (hunk 8 above)
Comment (alex) [inline docs/hazmat/backends/interfaces.rst]: PEM is an abbrevaitation that is more well known than the expanded form
(and also the expanded form is basically meaningless), I don't think that's
true of CSR. That said, it /is/ a very long name.

On Sun, Mar 15, 2015 at 5:10 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/hazmat/backends/interfaces.rst
> <https://github.com/pyca/cryptography/pull/1702#discussion_r26452379>:
>
> > @@ -509,3 +509,12 @@ A specific ``backend`` may provide one or more of these interfaces.
> >          :param bytes data: DER formatted certificate data.
> >
> >          :returns: An instance of :class:`~cryptography.x509.Certificate`.
> > +
> > +    .. method:: load_pem_x509_csr(data)
> > +
> > +        .. versionadded:: 0.9
> > +
> > +        :param bytes data: PEM formatted certificate request data.
>
> I discussed this with glyph a bit on IRC and it was his contention that
> load_pem_x509_csr is better than load_pem_x509_certificate_signing_request
> since pem is already an abbreviation. I am leaning towards leaving it as
> csr here but could be persuaded otherwise.
> load_pem_x509_certificate_signing_request is a pretty long name though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1702/files#r26452379>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: docs/x509.rst (hunk 4 above)
Comment (alex) [inline docs/x509.rst]: Hmm, yes I think so.

On Sun, Mar 15, 2015 at 5:11 PM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509.rst
> <https://github.com/pyca/cryptography/pull/1702#discussion_r26452382>:
>
> > +    bG8uY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAqhZx+Mo9VRd9
> > +    vsnWWa6NBCws21rZ0+1B/JGgB4hDsZS7iDE4Bj5z4idheFRtl8bBbdjPknq7BfoF
> > +    8v15Zq/Zv7i2xMSDL+LUrTBZezRd4bRTGqCm6YJ5EYkhqdcqeZleHCFImguHoq1J
> > +    Fh0+kObQrTHXw3ZP57a3o1IvyIUA3nNoCBL0QQhwBXaDXOojMKNR+bqB5ve8GS1y
> > +    Elr0AM/+cJsfaIahNQUgFKx3Eu3GeEOMKYOAG1lycgdQdmTUybLrT3U7vkClTseM
> > +    xHg1r5En7ALjONIhqRuq3rddYahrP8HXozb3zUy3cJ7P6IeaosuvNzvMXOX9P6HD
> > +    Ha9urDAJ1wIDAQABoDUwMwYJKoZIhvcNAQkOMSYwJDAiBgNVHREEGzAZggl3b3Js
> > +    ZC5jb22CDHdoYXRldmVyLmNvbTANBgkqhkiG9w0BAQUFAAOCAQEAS4Ro6h+z52SK
> > +    YSLCYARpnEu/rmh4jdqndt8naqcNb6uLx9mlKZ2W9on9XDjnSdQD9q+ZP5aZfESw
> > +    R0+rJhW9ZrNa/g1pt6M24ihclHYDAxYMWxT1z/TXXGM3TmZZ6gfYlNE1kkBuODHa
> > +    UYsR/1Ht1E1EsmmUimt2n+zQR2K8T9Coa+boaUW/GsTEuz1aaJAkj5ZvTDiIhRG4
> > +    AOCqFZOLAQmCCNgJnnspD9hDz/Ons085LF5wnYjN4/Nsk5tS6AGs3xjZ3jPoOGGn
> > +    82WQ9m4dBGoVDZXsobVTaN592JEYwN5iu72zRn7Einb4V4H5y3yD2dD4yWPlt4pk
> > +    5wFkeYsZEA==
> > +    -----END CERTIFICATE REQUEST-----
> > +    """.strip()
>
> Should I move the certificate PEM setup to live with it as part of this PR?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1702/files#r26452382>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: docs/x509.rst [hunk 9]
```diff
@@ -77,6 +100,36 @@ Loading Certificates
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. function:: load_pem_x509_csr(data, backend)
+
+    .. versionadded:: 0.9
+
+    Deserialize a certificate signing request (CSR) from PEM encoded data. PEM
+    requests are base64 decoded and have delimiters that look like
+    ``-----BEGIN CERTIFICATE REQUEST-----``. This is also known as PKCS#10
+    format.
```
Comment (alex) [inline docs/x509.rst]: I think this sentence reads a bit better as "This format is also known as PKCS#10"

**Comment by alex** [comment]
Tests look good.

**Comment by reaperhulk** [comment]
retest this please

**Comment by alex** [comment]
coverage.

**Comment by public** [comment]
LGTM

File: src/cryptography/hazmat/backends/openssl/x509.py [hunk 10]
```diff
@@ -187,3 +187,31 @@ def extensions(self):
             seen_oids.add(oid)
 
         return x509.Extensions(extensions)
+
```
Comment (alex) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Need an additional blank line here I believe.

Comment (reaperhulk) [inline src/cryptography/hazmat/backends/openssl/x509.py]: Done

**Comment by alex** [comment]
oof, merge conflict :-(
